test/gateway/hooks.js

```diff
@@ -178,15 +178,15 @@ test('gateway - hooks', async (t) => {
   app.graphql.addHook('preValidation', async function (schema, document, context) {
     await immediate()
     t.type(schema, GraphQLSchema)
-    t.deepEqual(document, parse(query))
+    t.type(document, 'object')
```

Comment:
why this change? The previous assertion wanted to measure (deep) equality with the document and query.

Reply:
I found that occasionally the GraphQL `parse` function times out when processing a query, leading to timeouts in the tests. This only occurred on the gateway code, so I changed this to a type assertion as we test this deep equality in other hook tests and therefore would be covered if anything changed in the hook runners.

As the query is a fixed string, I cannot think of a reason why this might be non-deterministic - other than issues in the `deepEqual` assertion or `parse` in some edge-case situations. I can take a deeper look at why the parse function is timing out occasionally, as I agree that ideally this should be retained if possible.

Reply:
that would be amazing, thanks

Reply:
cool, no problem - looking into it now! :)

Reply:
Good and bad news, I'm unable to replicate the hooks test error after ~15 CI runs so I will revert these changes :) However, I have found another test error in pollingInterval which occurs once in every 5 CI runs, so I will have a look at this as well!

Reply:
Hi! Firstly, apologies for the delayed response, it's been quite a slow process debugging these intermittent tests as they only occur every 5 CI runs or so, which adds up to ~30min between each failed test which makes it hard to verify that my change has actually worked!

Quick update, I've got some code ready to push which tidies up some of the `t.tearDown` calls as a lot of them don't return the promiselike `.close()` function, e.g.

```js
t.tearDown(() => {
  app.close()
})
```
I've changed them in this file to handle the promise behaviour correctly by returning directly or awaiting in order within an async function. After implementing this change in just one of the test files, it seems to reduce the likelihood of the final intermittent error - I am fairly sure about this as I have just run the CI on my fork 20 times with no failures at all! 

For the pollingInterval test that is occasionally failing (this is the only problem test left), it seems like it is exiting the test early after the 8th assertion:

```txt
# Subtest: Polling schemas (subscriptions should be handled)
        ok 1 - should be equal
        ok 2 - should be equal
        ok 3 - should be equal
        ok 4 - should be equal
        ok 5 - should be equivalent
        ok 6 - should be equivalent
        ok 7 - should be equal
        ok 8 - should be equal
    # test count(5) != plan(null)
    # failed 1 of 5 tests
        # test count(8) != plan(null)
        # failed 1 of 8 tests
```

Somewhere after this [line](https://github.com/mercurius-js/mercurius/blob/master/test/gateway/pollingInterval.js#L947) and the next assertion I suspect one of the calls is exiting prematurely. If you have any suggestions as to why this particular test might be exiting every 20 CI runs or so on MacOS/Windows, I would be very grateful!

Finally, a quick scan of the rest of the tests shows a fair few are still not awaiting or returning the `app.close()` function, so wondered if you wanted me to sort this out as well (very happy to!)? It's worth noting that I think it would be a fairly sizeable change (39 results in 9 files) and maybe a bit of refactoring to make sure the gateways are closed before the federated services!

Reply:
The refactoring would be quite welcomed, thanks!